docs: Tutorial that encourages interactive onboarding first (#1077)
windmilleng/tilt: docs/_build/html/_sources/tiltfile_concepts.md.txt

# Tiltfile Concepts

This doc describes concepts in the Tiltfile, expanding on the [Tutorial](tutorial.html). Unlike the [API Reference](api.html), it groups functions by themes and explains why you'd choose to use a function.

## Execution Model
`Tiltfile`s are written in [Starlark](https://github.com/bazelbuild/starlark), a dialect of Python. Tilt executes the `Tiltfile` on startup.

> Functions like `k8s_yaml` and `docker_build` register information. At the end of the execution, Tilt uses the resulting configuration. In addition to the final configuration, Tilt records file accesses. Tilt listens to these files, and reexecutes when one changes (but not on every source file change).

Reviewer: ```suggestion
Functions like `k8s_yaml` and `docker_build` register information. At the end of the execution, Tilt uses the resulting configuration. In addition to the final configuration, Tilt records file accesses; Tilt watches these files, and reexecutes when one changes (but not on every source file change).
```

> Because Tilt uses a programming language, you can configure it with familiar constructs like loops, functions, arrays, etc. This makes Tilt more extensible than a configuration that requires hard-coding all possible options up-front.

Reviewer: ```suggestion
Because your Tiltfile is written in a programming language, you can configure it with familiar constructs like loops, functions, arrays, etc. This makes Tilt more extensible than a configuration that requires hard-coding all possible options up-front.
```

Author: WDYT?

## Deploy
The first function in a `Tiltfile` is generally a call to `k8s_yaml`. You can call `k8s_yaml` in a variety of ways, depending on how your project organizes or generates yaml. Let's look at some alternatives:

```python
# one static yaml file
k8s_yaml('app.yaml')

# multiple yaml files in one call
k8s_yaml(['foo.yaml', 'bar.yaml'])

# multiple yaml files in multiple calls
k8s_yaml('baz.yaml')
k8s_yaml('quux.yaml')

# call out to a built-in tool
k8s_yaml(kustomize('config_dir')) # built-in support for popular tools
k8s_yaml(helm('chart_dir'))
```

Tilt has built-in functions to generate kubernetes yaml with `kustomize` or `helm`. (If you think we're overlooking a popular tool, let us know so we can add it.)

## Custom Commands

> If your project uses a custom tool to generate kubernetes yaml, you can still use Tilt. You don't have to wait for us to add support or fork Tilt and implement it yourself. Use the `local` function:

Reviewer: ```suggestion
If your project uses a custom tool to generate kubernetes yaml, you can still use Tilt. You don't have to wait for us to add support or fork Tilt and implement it yourself; just use the `local` function:
```

Author: WDYT?

The `docker_build` function aims to support most usages of docker. Here's a cheat-sheet that maps docker command lines to a `docker_build` call:

```python
# docker build -t companyname/frontend ./frontend
docker_build("companyname/frontend", "frontend")

# docker build -t companyname/frontend -f frontend/Dockerfile.dev frontend
docker_build("companyname/frontend", "frontend", dockerfile="frontend/Dockerfile.dev")

# docker build -t companyname/frontend --build-arg target=local frontend
docker_build("companyname/frontend", "frontend", build_args={"target": "local"})
```

You can combine multiple optional arguments.

## Resources
Tilt's UI makes it easier to find errors by grouping related status and output. E.g., when you edit a file, you want to know what error it caused, whether it's an error at build-time, deploy-time, or run-time. Tilt calls these groupings "Resources". Each Resource has a line in the UI that can be expanded and investigated.

Tilt generates these groups after executing your `Tiltfile`. We're actively working on how to group in ways that make the most intuitive sense, so the specific algorithm is in-flux. We'll expand this paragraph when it's more settled.

> You can configure a resource with a call to `k8s_resource`. Today, the only relevant configuration is the argument `port_forwards`. Tilt supports a few alternatives:

Reviewer: worth it to mention how we currently name resources/that you can rename them with this func?

Author: I think that's very likely to change, and if people need that we can help them on an ad hoc basis.
